internal/gojs/syscall.go

```diff
@@ -163,175 +138,150 @@ func valueSet(ctx context.Context, mod api.Module, sp []uint64) {
 // ValueDelete is stubbed as it isn't used in Go's main source tree.
 //
 // See https://github.com/golang/go/blob/go1.19/src/syscall/js/js.go#L321
-var ValueDelete = stubFunction(valueDeleteName)
+var ValueDelete = goarch.StubFunction(valueDeleteName)
 
 // ValueIndex implements js.valueIndex, which is used to load a js.Value property
 // by index, e.g. `v.Index(0)`. Notably, this is used by js.handleEvent to read
 // event arguments
 //
 // See https://github.com/golang/go/blob/go1.19/src/syscall/js/js.go#L334
 // and https://github.com/golang/go/blob/go1.19/misc/wasm/wasm_exec.js#L331-L334
-var ValueIndex = newSPFunc(valueIndexName, valueIndex)
-
-func valueIndex(ctx context.Context, mod api.Module, sp []uint64) {
-	mem := mod.Memory()
-
-	// Read (param + result count) * 8 memory starting at SP+8
-	stack := mustRead(mem, "sp", uint32(sp[0]+8), 24)
+var ValueIndex = goos.NewFunc(valueIndexName, valueIndex,
+	[]string{"v", "i"},
+	[]string{"r"},
+)
 
-	vRef := le.Uint64(stack)
-	i := le.Uint32(stack[8:])
+func valueIndex(ctx context.Context, _ api.Module, stack goos.Stack) {
+	v := stack.ParamVal(ctx, 0, loadValue)
+	i := stack.ParamUint32(1)
 
-	v := loadValue(ctx, ref(vRef))
 	result := v.(*objectArray).slice[i]
 
-	ref := storeRef(ctx, result)
-
-	// Write the results to memory at positions after the parameters.
-	le.PutUint64(stack[16:], ref)
+	r := storeRef(ctx, result)
+	stack.SetResultRef(0, r)
 }
 
 // ValueSetIndex is stubbed as it is only used for js.ValueOf when the input is
 // []interface{}, which doesn't appear to occur in Go's source tree.
 //
 // See https://github.com/golang/go/blob/go1.19/src/syscall/js/js.go#L348
-var ValueSetIndex = stubFunction(valueSetIndexName)
+var ValueSetIndex = goarch.StubFunction(valueSetIndexName)
 
 // ValueCall implements js.valueCall, which is used to call a js.Value function
 // by name, e.g. `document.Call("createElement", "div")`.
 //
 // See https://github.com/golang/go/blob/go1.19/src/syscall/js/js.go#L394
 // and https://github.com/golang/go/blob/go1.19/misc/wasm/wasm_exec.js#L343-L358
-var ValueCall = newSPFunc(valueCallName, valueCall)
+var ValueCall = goos.NewFunc(valueCallName, valueCall,
```

Comment:
this function actually implements most syscalls as file i/o is mapped at the JS layer in GOARCH=wasm/GOOS=js.

In order to do logging, we need to see the parameters. We now, have natural visibility except the expected names of the args. Ex. the args to "mkdir". Adding those will be in a follow-up PR as it is too much to do right now. This PR doesn't change any behavior, just re-organizes the code so that the abstractions are separated and also the packages that handle them are decoupled.